### PR TITLE
os/bluestore: refactor ExtentMap::update to avoid preceeding db updat…

### PR DESCRIPTION
**Pull request by ifed01**
…e if reshard takes place.

update method might add some 'set' ops to the transaction prior to detection that resharding is needed.
No need to apply these ops after reshard takes place - correspoding records to be removed anyway.

Signed-off-by: Igor Fedotov <ifed@mirantis.com>

### CODE REVIEW

**Comment by ifed01** [comment]
@liewegas  fixed and rebased

File: src/os/bluestore/BlueStore.cc [hunk 1]
```diff
@@ -1596,7 +1596,30 @@ bool BlueStore::ExtentMap::update(Onode *o, KeyValueDB::Transaction t,
     }
     // will persist in the onode key, see below
   } else {
+
+    struct dirty_shard_t{
+    string* key = nullptr;
+    bufferlist bl;
```
Comment (liewegas) [inline src/os/bluestore/BlueStore.cc]: If we put the bufferlist dirty_bl in Shard, we can avoid this structure entirely, right?

File: src/os/bluestore/BlueStore.cc [hunk 2]
```diff
@@ -1596,7 +1596,30 @@ bool BlueStore::ExtentMap::update(Onode *o, KeyValueDB::Transaction t,
...
+
+  typedef boost::intrusive::list<
+    dirty_shard_t,
+    boost::intrusive::member_hook<
+    dirty_shard_t,
+    boost::intrusive::list_member_hook<>,
+    &dirty_shard_t::dirty_list_item> > dirty_shard_list_t;
+
+
+    vector<dirty_shard_t> encoded_shards;
```
Comment (liewegas) [inline src/os/bluestore/BlueStore.cc]: and this

File: src/os/bluestore/BlueStore.cc [hunk 3]
```diff
@@ -1607,8 +1630,8 @@ bool BlueStore::ExtentMap::update(Onode *o, KeyValueDB::Transaction t,
 	} else {
 	  endoff = n->offset;
 	}
-	bufferlist bl;
 	unsigned nn;
+        bufferlist& bl = encoded_shards[pos].bl;
```
Comment (liewegas) [inline src/os/bluestore/BlueStore.cc]: bufferlist& bl = p->dirty_bl;

File: src/os/bluestore/BlueStore.cc [hunk 4]
```diff
@@ -1630,10 +1653,19 @@ bool BlueStore::ExtentMap::update(Onode *o, KeyValueDB::Transaction t,
...
+        dirty_shards.push_back(encoded_shards[pos]);
 	p->dirty = false;
       }
       p = n;
+      ++pos;
+    }
+    //schedule DB update for dirty shards
+    auto it = dirty_shards.begin();
+    while( it != dirty_shards.end()) {
+      t->set(PREFIX_OBJ, *(it->key), it->bl);
```
Comment (liewegas) [inline src/os/bluestore/BlueStore.cc]: and here we'd clear dirty_bl again?

I guess this uses a bit more RAM...

Comment (ifed01) [inline src/os/bluestore/BlueStore.cc]: And a small bonus of my approach - we don't need to enumerate ALL the shards twice. I'm iterating over dirty blobs only at the final stage.

**Comment by liewegas** [comment]
Yeah, let's go with what you ahve.  Can you clean up whitespace, and 
remove the unneed intrusive member item in Shard?

**Comment by ifed01** [comment]
@liewegas - cleaned-up

**Pull request merged**